documentation/security-architecture.md

```diff
@@ -0,0 +1,72 @@
+
+```bash
+# set a variable with a secret
+{ VAR='secret_value' ; } 2>>/dev/$SECRET_OUTPUT_DEV
+
+# use the variable with the secret
+{ COMMAND $SECRET_ARGUMENT ; } 2>>/dev/$SECRET_OUTPUT_DEV
+```
+
+### Protecting Against Code Injections
+
+*NOTE: This is our plan, the implementation will follow with the next release. This section will be updated accordingly.*
+
+ReaR is written in the Bash scripting language, the configuration is actually Bash script code and the ReaR system architecture is designed for easily extending ReaR functionality via dropping shell scripts into the ReaR script directories. Therefore the ReaR code is by nature more open than close, which is - from a security perspective - a trade-off for convenience and extensibility over security.
+
+We differentiate between the following scenarios for potential code injections:
+
+1. Reading data
+   1. from trusted operating system configuration files like `/etc/os-release`
+   2. from untrusted configuration files, for example provided by a backup tool that ReaR integrates with
+2. Reading ReaR scripts
+   1. when ReaR is installed
+   2. when ReaR is running in portable mode (`--portable`) or from source checkout
+3. Reading ReaR configuration files in `/etc/rear` or provided via the `-C` option
+
+To mitigate the risk of code injections, the following checks are implemented:
+
+* Trusted operating system configuration files defined as "shell compatible" are read with `source` as we must also support the different quoting styles that can be used there. This is considered safe, as the operating system configuration files are under the control of the operating system vendor and are not writable by unprivileged users.
+* Untrusted configuration files are never read with `source`. Instead, we use `grep` and other tools to extract the required information. This is considered safe, as the configuration files are not executed as code.
+* When ReaR is installed and not run from source or in portable mode, then we validate that ReaR scripts and configuration files are owned by `root` and are not writable by unprivileged users. This is considered safe, as the ReaR scripts are under the control of the ReaR package and are not writable by unprivileged users.
+
+Unfortunately, when running ReaR from a source checkout or in portable mode, we cannot implement further safeguards without compromising on core ReaR functionality. This is a trade-off between security and usability in favour of ReaR doing its designated job. If you run ReaR from a source checkout or in portable mode, you must take extra care and ensure that the ReaR scripts and configuration files are protected accordingly.
+
+The ReaR configuration is executable Bash code and this is an important feature to facilitate dynamic configuration. Therefore we cannot implement further checks or limits on the ReaR configuration files, but rely on the ReaR user to protect the configuration files accordingly.
```

Comment:
We could restrict the configuration language to be static (in particular no command substitution like `( ... )` ), but unfortunately if we wanted this to be enforced via the restricted shell mode (`set -r`), there would be a conflict with the recommended `2>>/dev/$SECRET_OUTPUT_DEV` (restricted shell does not allow redirections).

Reply:
I think at this stage we shouldn't introduce limitations that will render existing configurations incompatible.